cloud/aws/actuators/machine/actuator.go

```diff
@@ -104,6 +104,71 @@ func (a *Actuator) removeStoppedMachine(machine *clusterv1.Machine, client awscl
 	return TerminateInstances(client, instances, mLog)
 }
 
+func buildEc2Filters(inputFilters []providerconfigv1.Filter) []*ec2.Filter{
+	filters := make([]*ec2.Filter, len(inputFilters))
+	for i, f := range inputFilters {
+		values := make([]*string, len(f.Values))
+		for j, v := range f.Values {
+			values[j] = aws.String(v)
+		}
+		filters[i] = &ec2.Filter{
+			Name:   aws.String(f.Name),
+			Values: values,
+		}
+	}
+	return filters
+}
+
+func getSecurityGroupsIDs(securityGroups []providerconfigv1.AWSResourceReference, client awsclient.Client, mLog log.FieldLogger) ([]*string, error) {
+	var securityGroupIDs []*string
+	for _, g := range securityGroups {
+		// ID has priority
+		if g.ID != nil {
+			securityGroupIDs = append(securityGroupIDs, g.ID)
+		} else if g.Filters != nil {
+			mLog.Debug("Describing security groups based on filters")
+			// Get groups based on filters
+			describeSecurityGroupsRequest := ec2.DescribeSecurityGroupsInput{
+				Filters: buildEc2Filters(g.Filters),
+			}
+			describeSecurityGroupsResult, err := client.DescribeSecurityGroups(&describeSecurityGroupsRequest)
+			if err != nil {
+				mLog.Errorf("error describing security groups: %v", err)
+				return nil, fmt.Errorf("error describing security groups: %v", err)
+			}
+			for _, g := range describeSecurityGroupsResult.SecurityGroups {
+				groupID := *g.GroupId
+				securityGroupIDs = append(securityGroupIDs, &groupID)
+			}
+		}
+	}
+	return securityGroupIDs, nil
+}
+
+func getSubnetID(subnet providerconfigv1.AWSResourceReference, client awsclient.Client, mLog log.FieldLogger) (*string, error) {
+	var subnetIDs []*string
+	// ID has priority
+	if subnet.ID != nil {
+		subnetIDs = append(subnetIDs, subnet.ID)
+	} else {
+		mLog.Debug("Describing subnets based on filters")
+		describeSubnetRequest := ec2.DescribeSubnetsInput{
+			Filters: buildEc2Filters(subnet.Filters),
+		}
+		describeSubnetResult, err := client.DescribeSubnets(&describeSubnetRequest)
+		if err != nil {
+			mLog.Errorf("error describing security groups: %v", err)
+			return nil, fmt.Errorf("error describing security groups: %v", err)
+		}
+		for _, n := range describeSubnetResult.Subnets {
+			subnetID := *n.SubnetId
+			subnetIDs = append(subnetIDs, &subnetID)
```

Comment:
Looks like you could just return here on the first one and stop iterating the loop if we're just looking to return the first.

Should we warn if multiple results were returned?

Reply:
thanks! made the function able to return multiple ids, then upper level function decideds what to do and warns